### ISSUE TITLE
Disposable Interface

### ISSUE THREAD

**Issue by srikrsna** [flutter/flutter#86615]
## Use case

As of this writing, a lot classes in the SDK have a method defined on them with the signature `void dispose()`. Some of the examples are,

* `TabController`
* `TextEditingController`
* `PageController`
* `ScrollController`
* `State`
* `ChangeNotifier`

Let's call them disposables.

Right now I cannot write functions/methods/classes that can accept a disposable.

## Proposal

I propose that a new abstract class definition be added to Flutter with the following definition,
```dart
abstract class Disposable {
  void dispose();
}
```

This abstract class should be implemented within the SDK where ever it already is implicitly satisfied. Eg:

```dart
class ChangeNotifier implements Listenable, Disposable 
```

* No breaking changes.
* No new API (no new methods/functions only marking existing classes).

**Comment by alexeyinkin** [comment]
- Duplicate of https://github.com/dart-lang/sdk/issues/43490 ?

**Comment by agrapine** [comment]
@alexeyinkin i don't think this is a duplicate. looks to me 'Disposable' is really a Flutter contract more than a language feature.

**Comment by alexeyinkin** [comment]
If we make it a Flutter feature, we leave the pure-Dart `StreamController` and `StreamSubscription` behind. So we still will have disposable classes not covered by this interface, which defeats the point of introducing it.

**Comment by agrapine** [comment]
@alexeyinkin not sure that is the point. StreamController & StreamSubscription have an async `cancel`. not quite the same thing. AsyncDisposable would be required for them as they need the caller to be also colored as `async`.

**Comment by alexeyinkin** [comment]
In that issue, `cancel()` vs `dispose()` is being questioned: https://github.com/dart-lang/sdk/issues/43490#issuecomment-695754533
Those seem to be different names for the same concept. You have something, want to get rid of it, call that thing, simple.

Also, I think that the decisions on `Disposable` and `AsyncDisposable` should be coordinated wherever each of them ends up. That's the reason to mark it duplicate.

And we should be careful putting generic things that are valid in pure Dart in Flutter. For instance, missing `Listenable` in Dart is really inconvenient, see
- https://github.com/flutter/flutter/issues/136155

**Comment by agrapine** [comment]
I agree with you @alexeyinkin. it would definetly be quite nice to have this Disposable contract in pure Dart
